paddle/fluid/operators/softmax_with_cross_entropy_op.cu

```diff
@@ -53,8 +55,157 @@ __global__ void SoftCrossEntropyGradientKernel(T* logit_grad,
     logit_grad[ids] = loss_grad[row_ids] * (logit_grad[ids] - labels[ids]);
   }
 }
+
 }  // namespace
 
+__device__ __forceinline__ float real_exp(float x) { return expf(x); }
+__device__ __forceinline__ double real_exp(double x) { return exp(x); }
+__device__ __forceinline__ float real_log(float x) {
+  return math::TolerableValue<float>()(logf(x));
+}
+__device__ __forceinline__ double real_log(double x) {
+  return math::TolerableValue<double>()(log(x));
+}
+
+// There are 3 kinds of reduce algorithm:
+// BLOCK_REDUCE_RAKING_COMMUTATIVE_ONLY
+// BLOCK_REDUCE_RAKING
+// BLOCK_REDUCE_WARP_REDUCTIONS (default)
+template <typename T, int BlockSize>
+using BlockReduce =
+    cub::BlockReduce<T, BlockSize /*, cub::BLOCK_REDUCE_WARP_REDUCTIONS*/>;
+
+template <typename T, int BlockSize>
+using BlockReduceTempStorage = typename BlockReduce<T, BlockSize>::TempStorage;
+
+// Make sure that BlockSize <= feature_size
+template <typename T, int BlockSize>
+__global__ void RowReductionForMax(const T* logits_data, T* max_data,
+                                   int feature_size) {
+  __shared__ BlockReduceTempStorage<T, BlockSize> temp_storage;
```

Comment:
The BlockSize is just needed by this type. What if we set BlockSize to MaxBlockSize?